classes/observers/traits/UsageEvent.inc.php

```diff
@@ -0,0 +1,164 @@
+    public int $assocType;
+
+    /** Canonical URL for the pub object */
+    public string $canonicalUrl;
+
+    public Request $request;
+
+    public Context $context;
+
+    public ?Submission $submission;
+
+    /** Representation (galley or publication format) */
+    public ?Representation $representation;
+
+    public ?SubmissionFile $submissionFile;
+
+    /** Application's complete version string */
+    public string $version;
+
+    /**
+     * Create a new usage event instance.
+     */
+    protected function constructUsageEvent(int $assocType, Context $context, Submission $submission = null, Representation $representation = null, SubmissionFile $submissionFile = null)
+    {
+        $application = Application::get();
+        $this->request = $application->getRequest();
+
+        $canonicalUrlPage = $canonicalUrlOp = null;
+        $canonicalUrlParams = [];
+
+        switch ($assocType) {
+            case Application::ASSOC_TYPE_SUBMISSION_FILE_COUNTER_OTHER:
+            case Application::ASSOC_TYPE_SUBMISSION_FILE:
+                $canonicalUrlOp = 'download';
+                $canonicalUrlParams = [$submission->getId()];
+                $router = $this->request->getRouter(); /** @var PageRouter $router */
+                $op = $router->getRequestedOp($this->request);
+                $args = $router->getRequestedArgs($this->request);
+                if ($op == 'download' && count($args) > 1) {
+                    if ($args[1] == 'version' && count($args) == 5) {
+                        $publicationId = (int) $args[2];
+                        $canonicalUrlParams[] = 'version';
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                $canonicalUrlParams[] = $representation->getId();
+                $canonicalUrlParams[] = $submissionFile->getId();
+                break;
+            case Application::ASSOC_TYPE_SUBMISSION:
+                $canonicalUrlOp = 'view';
+                if ($application->getName() == 'omp') {
+                    $canonicalUrlOp = 'book';
+                }
+                $canonicalUrlParams = [$submission->getId()];
+                $router = $this->request->getRouter(); /** @var PageRouter $router */
+                $op = $router->getRequestedOp($this->request);
+                $args = $router->getRequestedArgs($this->request);
+                if ($op == $canonicalUrlOp && count($args) > 1) {
+                    if ($args[1] == 'version' && count($args) == 3) {
+                        $publicationId = (int) $args[2];
+                        $canonicalUrlParams[] = 'version';
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                break;
+            case Application::getContextAssocType():
+                $canonicalUrlOp = '';
+                $canonicalUrlPage = 'index';
+                break;
+        }
```

Comment:
This should throw an exception in the default case.

Reply:
I needed to reorganize the code a little bit for this...